### ISSUE TITLE
User mention automatically makes this user a reviewer 

### ISSUE THREAD

**Issue by Seybo** [Reviewable/Reviewable#565]
A person becomes a reviewer of the whole PR if he was just mentioned in a single comment. Would be great if in this case this person would only have a one _to repy_ notification

**Comment by Seybo** [comment]
hi @pkaminski 

> If they have previously marked files as reviewed (even in draft), then they're a reviewer.

do you mean you have to make this person a reviewer (because of the github restrictions)? Because like i said earlier this is not right to my mind. If someone just marked a file in a draft and not published it then he is not a reviewer yet. Noone saw his review and maybe will never see it (if he won't publish). So a reviewer is someone who did a review. Which includes publishing it
 
> If they are a requested reviewer or have posted a GitHub review with either APPROVED or CHANGES_REQUESTED state, then they're a reviewer.

i'm not a pro github user but what about posting review without any state or with any other state?

> At this point, if they're the author of the PR they're not a reviewer.

I think the author of the PR should never be a reviewer of his own PR. It just doesn't make sense, does it?

> If they are an assignee, then they're a reviewer.

+

> ... here things get tricky. We have basically no info about the user and need to decide whether to treat them as a reviewer or not. We could say that if they're mentioned (and none of the above apply) then they're not a reviewer. But what if somebody (e.g., the author or another reviewer) just mentioned the person to attract their attention to some point during the review and they didn't get started yet? 

hmmm... so what is the problem here? If someone was just mentioned he is not a reviewer. If this person is already a reviewer he remains a reviewer. Long story short: mentioning shouldn't make any difference to the reviewer state of a mentioned user

> Or what if the team doesn't use assignees or requested reviewers, and instead @-mentions the people it would like to review the PR?

No problem here i think. If someone was pinged with mentioning then it's up to him: whether to become a reviewer or not. And becoming a reviewer means publishing a review (well if it's not possible then at least starting a draft review)

> I don't think there's enough data to make a decision about mentioned users in step 5 without a high error rate either way. Thoughts and ideas welcome...

Probably i'm not getting what confuses you here. To me everything seems absolutely clear and safe

**Comment by pkaminski** [comment]
I think the bit you're missing is that it's important to guess whether a person is a reviewer _before_ they've started the review, or at the latest _as soon as_ they've started the review.  Otherwise, their first review round will be unguided: discussions won't show up as unreplied, files won't show up as to-review, etc.  Thus identifying somebody as "just mentioned, not a reviewer" is a trade-off in false positives and negatives, with insufficient information to tell up front.  I'm not yet sure where the balance lies.

Also, self-reviews are a pretty common workflow, so yes the PR author does sometimes need to be considered a reviewer.

**Comment by pkaminski** [comment]
I've now spent over 2 hours thinking about this and I don't believe it's possible to distinguish between mentioned-for-spot-advice vs mentioned-for-full-review with better than 50/50 reliability, in which case I'd rather bias towards reviewers.

However, I'd like to go back to your original premise because I'm wondering if there's a different kind of bug there instead:  did you say that when you were mentioned and visited the review, _all_ the discussions were flagged as "to reply"?  The intention (if you're not the PR author) is that only discussions you were mentioned in, the top-level discussion, and discussions initiated by the PR author that don't have any responses yet should count as unreplied.  Since the last set is usually empty, that _ought_ to mean only the discussion that mentions you and the top-level thread are flagged.  Do you see a different behavior?

**Comment by Seybo** [comment]
> did you say that when you were mentioned and visited the review, all the discussions were flagged as "to reply"?

yeah. As far as remember. I can't check it now. Need to wait until i will be mentioned in a situation like this againg

> and discussions initiated by the PR author that don't have any responses yet should count as unreplied.
 
maybe that was the case. All the discussions didn't yet have any responses at that time

**Comment by Seybo** [comment]
> I think the bit you're missing is that it's important to guess whether a person is a reviewer before they've started the review, or at the latest as soon as they've started the review. Otherwise, their first review round will be unguided: discussions won't show up as unreplied, files won't show up as to-review, etc. 

could you please elaborate? What exactly is the first round (when does it start)? And why it will be unguided in this case?

**Comment by pkaminski** [comment]
> maybe that was the case. All the discussions didn't yet have any responses at that time

Were those discussions started by the PR author, or some other reviewer?  If the latter, they shouldn't have come up as unreplied for you.  Now that you know what to look for let's see what the state is next time.

> could you please elaborate? What exactly is the first round (when does it start)? And why it will be unguided in this case?

You were suggesting that only people who submitted a review should be treated as reviewers.  But they already need to be treated as reviewers in the process of producing this first review, before it's submitted (this is what I call the first round).  In principle, this review process starts as soon as they land on the page -- there's no special UI affordance for saying "I'm starting a review now".  So when a potential reviewer lands on the review page for the first time, we already need to show them the full "to review" and "unreplied" counters so they can navigate easily.  And unfortunately, being mentioned is not a strong indication of whether the mentionee will play the role of reviewer or not.

**Comment by pkaminski** [comment]
> by the PR author. But that also doesn't make sense, does it?

That's unusual and explains why everything showed up as unreplied for you.  If the PR author starting lots of discussions happens often in your repos, can say more about your workflow?

Let me try another tack to explain why this issue is hard to resolve.  We agree that if somebody who intends to be a reviewer arrives at the page we should show them all discussions as unreplied and all files as to be reviewed.  On the other hand, if a, let's say, "mentioned consultant" arrives at the page, we should only show as unreplied the discussions they were mentioned in and not prompt them to review any files.  The former might be invoked as "@seybo, could you please review this PR?" and the latter as "@seybo, what do you think about this one thing X?".  But from the point of view of Reviewable, these are indistinguishable!  All we know is that @seybo was mentioned, but we can't divine the intent of the mention.

**Comment by pkaminski** [comment]
> that's where we don't agree actually :) If someone intends to be a reviewer then he should take an action first (i.e. comment something or mark a file as reviewed) and only then he should be treated as a reviewer. And the best way i guess would be to add a button somewhere: `Become a reviewer`

Ah, then I'm afraid we have an irresolvable disagreement here:  since the vast majority of people visiting a review page (other than the PR author) are going to be reviewers, it's not acceptable to me to add an extra button click to their workflow.  But I still really want to address the issue so at this point I've formulated short term and long term plans:

Short term, I intend to mark users who are mentioned in a discussion _other than the top-level thread_ as mentionees, and unless they take an affirmative action to treat them as non-reviewers.  I _think_ this is a reasonable trade-off, but I'll be monitoring for feedback and if the false positive rate ends up too high for my liking, I will revert the change.

Long term, I'm considering giving users explicit control over their role in a review.  A default role would still be prepopulated for them (based on all the semantics discussed so far) but they'd be able to override it.  This would make me more comfortable with a higher false positive rate above and, if I had to revert the 'mentionee' change I'd be inclined to give it another try.  (There are other use cases arguing for explicit role control as well, such as being able to have multiple "PR authors".)  Note that implementing this will have to wait for some (previously planned) major UI changes, though.
